SQL: Add BigDecimal support to JDBC

bpintea opened the pull request:
This PR implements the JDBC support for BigDecimals. This is limited however, since ES support for this type is limited.
Fetching BigDecimals works with no restrictions, since all ES native and SQL numeric (and boolean) types can be accommodated by a BigDecimal.
Setting a BigDecimal value for a parameter in a prepared statement is limited to what a double can encode.

Closes #43806

bpintea committed:
Introduce BigDecimal support to JDBC -- fetching

This commit adds support for the getBigDecimal() methods.

bpintea committed:
Allow BigDecimal params in double range

A prepared statement will now accept a BigDecimal parameter as a proxy
for a double, if the conversion is lossless.

bpintea commented on x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/PreparedStatementTestCase.java:
```diff
@@ -16,16 +17,12 @@
 public class PreparedStatementTestCase extends JdbcIntegrationTestCase {
 
     public void testSupportedTypes() throws Exception {
-        index("library", builder -> {
-            builder.field("name", "Don Quixote");
-            builder.field("page_count", 1072);
-        });
-
```
I've removed an index involvement mostly due to #56013. Using an index for this test doesn't seem to bring any particular value, but can revert if I'm missing anything.

bpintea committed:
Styling, line limit

- adjust unchanged line to newly enforced line limit

bpintea committed:
style check fix

Remove wrongly added chunk.

bpintea committed:
Merge remote-tracking branch 'upstream/master' into enh/jdbc_bigdecimal

elasticmachine commented:
Pinging @elastic/es-ql (:Query Languages/SQL)

costin commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/JdbcResultSet.java:
```diff
@@ -547,7 +550,9 @@ public InputStream getBinaryStream(int columnIndex) throws SQLException {
     @Deprecated
     public BigDecimal getBigDecimal(String columnLabel, int scale) throws SQLException {
-        throw new SQLFeatureNotSupportedException("BigDecimal not supported");
+        BigDecimal bd = getBigDecimal(columnLabel);
```
Why not delegate to the other `getBigDecimal` -> `return getBigDecimal(column(columnLabel), scale)` ?

bpintea replied:
Right, good point! Thanks.

costin reviewed (approved):
LGTM

matriv commented on x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/ResultSetTestCase.java:
```diff
@@ -791,7 +789,188 @@ public void testGettingInvalidFloat() throws Exception {
+
+        doWithQuery(SELECT_WILDCARD, (results) -> {
+            results.next();
+            for(Entry<String, Number> e : map.entrySet()) {
```
```suggestion
            for (Entry<String, Number> e : map.entrySet()) {
```

matriv commented on x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/ResultSetTestCase.java:
```diff
@@ -791,7 +789,188 @@ public void testGettingInvalidFloat() throws Exception {
+        put(Long.class, Types.BIGINT);
+        put(Float.class, Types.REAL);
+        put(Double.class, Types.DOUBLE);
+        // TODO: no half & scaled float testing
```
Why not adding those too?

bpintea replied:
It probably should be done indeed, but I was thinking to keep it focused on BigDecimal for now, since adding coverage for half- and scaled- floats would require extending the testing for the rest of the types too.

matriv reviewed (approved):
LGTM. Nice enhancements on the ResultSet testing!
Left a couple of comments.

bpintea committed:
Adress review comments

Remove code duplicate in getBigDecimal(String columnLabel, int scale).

bpintea committed:
Update x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/ResultSetTestCase.java

Co-authored-by: Marios Trivyzas <matriv@gmail.com>

bpintea committed:
Suppress failure due to use of deprecated APIs

Disable the deprecation-API usage failure, since this is used to
implement a deprecated API itself.

bpintea committed:
Merge branch 'enh/jdbc_bigdecimal' of github.com:bpintea/elasticsearch into enh/jdbc_bigdecimal

astefan reviewed (approved):
LGTM
